### ISSUE TITLE
SCC disconnecting mid-game

### ISSUE THREAD

**Issue by ibanman555** [kozec/sc-controller#615]
New user here to allow for all buttons of SC in games, and set up a simple Press/Double Press of menu button to toggle between Desktop and XBOX controls. Intermittent disconnection of the controller requires a force quit or reboot, even though everything appears to be on and emulating still. SC connected to RPI4, streaming Moonlight. I wish it was more stable, not sure what to do.

**Comment by kozec** [comment]
What to do is to turn off everything SCC-related and running `scc-daemon debug` in terminal. Then, once Intermittent disconnection happens, check and post error messages. We may be able to figure out what's wrong.

**Comment by ibanman555** [comment]
Ok thanks. I'll get on this...

**Comment by ibanman555** [comment]
Here's what I have from the last attempt to get a disconnection.  I tried to turn the controller off and on again, but it didn't reconnect. I then had to pull the usb dongle and reinsert for connection to re-establish:

`pi@raspberrypi:~ $ scc-daemon debug
D SCCDaemon     Starting SCCDaemon...
D SCCDaemon     Initializing drivers...
W SCCDaemon     Skipping disabled driver 'ds4drv'
W SCCDaemon     Skipping disabled driver 'fake'
W SCCDaemon     Skipping disabled driver 'remotepad'
W evdev         Failed to enable Evdev driver: 'python-evdev' package is missing.
D USB           Registered USB driver for 28de:1102
D USB           Registered USB driver for 28de:1142
W DevMon        Failed to load libbluetooth.so, bluetooth support will be incomplete
D Mapper        Creating virtual devices
D Mapper        Keyboard: <scc.uinput.Keyboard object at 0xb5e6d930>
D Mapper        Mouse:    <scc.uinput.Mouse object at 0xb5dcc9b0>
D Mapper        Gamepad:  <scc.uinput.UInput object at 0xb5dccad0>
D SCCDaemon     Created control socket /home/pi/.config/scc/daemon.socket
D SCCDaemon     Connected to XServer :0.0
D USB           USB device added: 28de:1142
D OSD DaemonCtrl    Connected to daemon, version 0.4.8
D OSD DaemonCtrl    Daemon is ready.
I SCCDaemon     Registered scc-osd-daemon
I OSD osd.daemon    Sucessfully registered as scc-osd-daemon
D SCDongle      Not requesting serial number for SC sc1
D SCDongle      Got wireless SC with serial sc1
D SCCDaemon     Assigned default_mapper to <SCWireless sc1>
D SCCDaemon     Turning gyrosensor ON
D SCCDaemon     Controller added: <SCWireless sc1>
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
I CemuHook      Created CemuHookUDP Motion Provider
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
I SCCDaemon     Loaded profile 'Desktop'
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
I SCCDaemon     Loaded profile 'Desktop'
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
I SCCDaemon     Loaded profile 'Desktop'
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
D SCCDaemon     Controller removed: <SCWireless sc1>
D USB           USB device added: 28de:1142
D OSD DaemonCtrl    Daemon is ready.
D SCDongle      Not requesting serial number for SC sc1
D SCDongle      Got wireless SC with serial sc1
D SCCDaemon     Assigned default_mapper to <SCWireless sc1>
D SCCDaemon     Turning gyrosensor ON
D SCCDaemon     Controller added: <SCWireless sc1>
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
I SCCDaemon     Loaded profile 'Desktop'
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
D SCCDaemon     Controller removed: <SCWireless sc1>
D USB           USB device added: 28de:1142
D OSD DaemonCtrl    Daemon is ready.
D SCDongle      Not requesting serial number for SC sc1
D SCDongle      Got wireless SC with serial sc1
D SCCDaemon     Assigned default_mapper to <SCWireless sc1>
D SCCDaemon     Turning gyrosensor ON
D SCCDaemon     Controller added: <SCWireless sc1>
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
I SCCDaemon     Loaded profile 'XBox Controller'

pi@raspberrypi:~ $ scc-daemon debug
D SCCDaemon     Starting SCCDaemon...
D SCCDaemon     Initializing drivers...
W SCCDaemon     Skipping disabled driver 'ds4drv'
W SCCDaemon     Skipping disabled driver 'fake'
W SCCDaemon     Skipping disabled driver 'remotepad'
W evdev         Failed to enable Evdev driver: 'python-evdev' package is missing.
D USB           Registered USB driver for 28de:1102
D USB           Registered USB driver for 28de:1142
W DevMon        Failed to load libbluetooth.so, bluetooth support will be incomplete
D Mapper        Creating virtual devices
D Mapper        Keyboard: <scc.uinput.Keyboard object at 0xb5e6d930>
D Mapper        Mouse:    <scc.uinput.Mouse object at 0xb5dcc9b0>
D Mapper        Gamepad:  <scc.uinput.UInput object at 0xb5dccad0>
D SCCDaemon     Created control socket /home/pi/.config/scc/daemon.socket
D SCCDaemon     Connected to XServer :0.0
D USB           USB device added: 28de:1142
D OSD DaemonCtrl    Connected to daemon, version 0.4.8
D OSD DaemonCtrl    Daemon is ready.
I SCCDaemon     Registered scc-osd-daemon
I OSD osd.daemon    Sucessfully registered as scc-osd-daemon
D SCDongle      Not requesting serial number for SC sc1
D SCDongle      Got wireless SC with serial sc1
D SCCDaemon     Assigned default_mapper to <SCWireless sc1>
D SCCDaemon     Turning gyrosensor ON
D SCCDaemon     Controller added: <SCWireless sc1>
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
I CemuHook      Created CemuHookUDP Motion Provider
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
I SCCDaemon     Loaded profile 'Desktop'
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'Desktop'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
I SCCDaemon     Loaded profile 'Desktop'
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
I SCCDaemon     Loaded profile 'Desktop'
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
D SCCDaemon     Controller removed: <SCWireless sc1>
D USB           USB device added: 28de:1142
D OSD DaemonCtrl    Daemon is ready.
D SCDongle      Not requesting serial number for SC sc1
D SCDongle      Got wireless SC with serial sc1
D SCCDaemon     Assigned default_mapper to <SCWireless sc1>
D SCCDaemon     Turning gyrosensor ON
D SCCDaemon     Controller added: <SCWireless sc1>
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/Desktop.sccprofile
I SCCDaemon     Loaded profile 'Desktop'
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
W Actions       Action ChangeProfileAction can't handle button press event
I SCCDaemon     Loaded profile 'XBox Controller'
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D OSD Config        Configuration saved
D OSD osd.daemon    Updated recent profile list
D SCCDaemon     Controller removed: <SCWireless sc1>
D USB           USB device added: 28de:1142
D OSD DaemonCtrl    Daemon is ready.
D SCDongle      Not requesting serial number for SC sc1
D SCDongle      Got wireless SC with serial sc1
D SCCDaemon     Assigned default_mapper to <SCWireless sc1>
D SCCDaemon     Turning gyrosensor ON
D SCCDaemon     Controller added: <SCWireless sc1>
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
W Actions       Action ChangeProfileAction can't handle button press event
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
I SCCDaemon     Loaded profile 'XBox Controller'


`

**Comment by kozec** [comment]
How sure you are it's not HW issue?

```
D SCCDaemon     Controller removed: <SCWireless sc1>
D USB           USB device added: 28de:1142
```

lines suggests dongle was disconnected and properly reconnected back, from SCC point of view. Since it's PI, 1st thing I'd check is whether it gets enough power and maybe `dmesg` may show something related.

**Comment by ibanman555** [comment]
> How sure you are it's not HW issue?
I am not very sure, so I will change USB port from the USB 3 to a USB 2 port. The dongle was extended from behind the Pi with a 10ft USB (I understand this is not ideal) extension for better reception, but I will use a much shorter cable and test.

Using SCC resulted in more disconnects and required physically pulling and reconnecting the dongle to reset versus directly connecting to Moonlight without SCC, which only requires power cycling the controller. So I was curious if anything preventing reconnection could be seen in the log. I'll try a few more things to see i I can eliminate this before I close.

**Comment by ibanman555** [comment]
Ok so I've done a few hours with and without SCC. Unfortunately the steam controller is dropping out and not reconnecting while using SCC, removing and reinserting USB dongle is the only way to reset. If I eliminate SCC the controller continues to function without dropping out.

**Comment by ibanman555** [comment]
![Screenshot_20210202-161938](https://user-images.githubusercontent.com/4698022/106664236-c6262d80-6572-11eb-9285-4f411d1ace0d.png)

**Comment by ibanman555** [comment]
I've been trying over and over, debug shows nothing. Controller just stops, locks whatever button/direction last pressed. Any Steam Controller vibrations cease. Sc-controller shows  "sc-0" is still connected in sc-controller, even after Steam Controller is physically turned off. If I turn it back on, it doesn't reconnect to sc-controller, or as in lizard mode. Pulling and reconnecting dongle will reset sc-controller properly, or cycling emulation will, as well. Exception in Thread below occured after cycling emulation.

```D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/pi/.config/scc/profiles/XBox Controller.sccprofile
D SCDongle      Turning off the controller...
D SCDongle      Turning off the controller...
D SCDongle      Turning off the controller...
D USB           Releasing devices...
D SCCDaemon     Controller removed: <SCWireless sc1>
Exception in thread Thread-3 (most likely raised during interpreter shutdown):pi@raspberrypi:~ $ ```

**Comment by ibanman555** [comment]
I believe I have found the solution to my specific  problem after a number of changes. From what I can understand my Pi was overclocked using
```over_voltage=6
arm_freq=2000
```
The purpose was to increase streaming performance, but clearly the OC'd Pi was pulling too much power away from the USB ports powering the Steam Controller dongle. I have decreased the OC settings to 
```over_voltage=4
arm_freq=1900
```
The disconnects seem to have been eliminated at this point. I still don't see why using the SC without sc-controller continued to function at these high OC settings, versus using it, but it points to an issue with sc-controller bugging out if there's a blip in the signal. None-the-less, I am glad to have possibly solved my particular issue and maybe can shed light on others using this setup, or improving your program.
